Fix imports for images with no info about MediaType.

tomob opened the pull request:
Signed-off-by: Tomasz Baranski <tbaransk@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

An image from a registry might have MediaType for its layer empty. This
PR adds code which tries to guess the MediaType basing on so-called
magic numbers aka file signatures.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

awels commented on pkg/image/transport.go:
```diff
@@ -129,6 +133,71 @@ func isDir(path string) bool {
+		head[0] == 0x1F && head[1] == 0x8B
+}
+
+func guessMediaType(ctx context.Context, src types.ImageSource, layer types.BlobInfo, cache types.BlobInfoCache) (string, error) {
```
This feels like it is doing the same as https://github.com/kubevirt/containerized-data-importer/blob/master/pkg/importer/format-readers.go#L227 more or less. But I can't come up with a good way of using that here of the top of my head.

tomob replied:
Oh, I didn't know about this code.

I've given it some thought. Maybe we could get rid of the whole `MediaType` checking and just use FormatReaders? If a layer can be read using FR, it's good enough for the import process. The declared edia type wouldn't matter.

What do you think?

awels replied:
Yeah I think that would be good. I just hate re-implementing code we have been using successfully for some time now. Also maybe some unit tests to verify it is working as expected?

tomob commented:
/retest

awels reviewed (approved):
/lgtm
/approve

awels commented:
looks like there are some failing tests now....

tomob commented:
> looks like there are some failing tests now....

This is weird... With `FormatReaders` reading is very sloooooow. But I can't see any reason for it.

tomob commented on pkg/importer/transport.go:
```diff
@@ -36,15 +34,11 @@ import (
 func commandTimeoutContext() (context.Context, context.CancelFunc) {
-	ctx := context.Background()
-	return context.WithTimeout(ctx, cancelTimeout)
+	return context.WithCancel(context.Background())
```
@awels I think we cannot say what a "generic" timeout for any situation should be. So I got rid of it; the importer will now patiently download images for as long as it takes.

tomob commented on tools/cdi-func-test-file-host-init/nginx.conf:
```diff
@@ -120,7 +120,7 @@ http {
             proxy_read_timeout                 900;
 
-            limit_rate 100k;
+            limit_rate 1M;
```
@awels It seems 100k (per second) was too harsh. The test was taking ages to run.

awels reviewed (commented):
One question, you mentioned with the FormatReaders it was much slower. How did we fix that? Hopefully not by just waiting longer. Or was the timeout causing a restart and that is why it was slower?

tomob commented:
> One question, you mentioned with the FormatReaders it was much slower. How did we fix that? Hopefully not by just waiting longer. Or was the timeout causing a restart and that is why it was slower?

No, I found the bug in the old code (the one I replaced) which made it skip some image layers (their media type was `...tar`, while it actually was gzipped, so the type should have been `....tar+gz`). After I fixed that both versions took about the same.

So, it's not that FormatReaders are slower – it's that my code that was too fast because it skipped layers it should process.

awels commented:
Okay so by switching to code we knew was working we fixed a bug before it became a bug. Nice

awels commented:
/lgtm

awels commented:
/cherrypick release-v1.23
